components/root/switch/switch.jsx

```diff
@@ -0,0 +1,321 @@
+const ErrorPage = makeAsyncComponent(loadErrorPage);
+const TermsOfService = makeAsyncComponent(loadTermsOfService);
+const LoginController = makeAsyncComponent(loadLoginController);
+const AdminConsole = makeAsyncComponent(loadAdminConsole);
+const LoggedIn = makeAsyncComponent(loadLoggedIn);
+const PasswordResetSendLink = makeAsyncComponent(loadPasswordResetSendLink);
+const PasswordResetForm = makeAsyncComponent(loadPasswordResetForm);
+const SignupController = makeAsyncComponent(loadSignupController);
+const SignupEmail = makeAsyncComponent(loadSignupEmail);
+const ShouldVerifyEmail = makeAsyncComponent(loadShouldVerifyEmail);
+const DoVerifyEmail = makeAsyncComponent(loadDoVerifyEmail);
+const ClaimController = makeAsyncComponent(loadClaimController);
+const HelpController = makeAsyncComponent(loadHelpController);
+const GetIosApp = makeAsyncComponent(loadGetIosApp);
+const GetAndroidApp = makeAsyncComponent(loadGetAndroidApp);
+const SelectTeam = makeAsyncComponent(loadSelectTeam);
+const Authorize = makeAsyncComponent(loadAuthorize);
+const Mfa = makeAsyncComponent(loadMfa);
+
+const LoggedInRoute = ({component: Component, ...rest}) => (
+    <Route
+        {...rest}
+        render={(props) => (
+            <LoggedIn {...props}>
+                <Component {...props}/>
+            </LoggedIn>
+        )}
+    />
+);
+
+export default class RootSwitch extends React.PureComponent {
+    static propTypes = {
+        currentUserId: PropTypes.string,
+        defaultRoute: PropTypes.string.isRequired,
+        noAccounts: PropTypes.bool.isRequired,
+        mfaRequired: PropTypes.bool.isRequired,
+        showTermsOfService: PropTypes.bool.isRequired,
+        iosDownloadLink: PropTypes.string,
+        androidDownloadLink: PropTypes.string,
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            currentUserIdSet: 0,
+            showTermsOfServiceSet: 0,
+        };
+    }
+
+    componentDidUpdate(prevProps) {
+        /* eslint-disable react/no-did-update-set-state */
+
+        // Record the time when the currentUserId is first set.
+        if (!prevProps.currentUserId && this.props.currentUserId) {
+            this.setState({
+                currentUserIdSet: Date.now(),
+            });
+        }
+
+        // Record the time when showTermsOfService is first set.
+        if (!prevProps.showTermsOfService && this.props.showTermsOfService) {
+            this.setState({
+                showTermsOfServiceSet: Date.now(),
+            });
+        }
+    }
+
+    shouldShowTermsOfService() {
+        // Honour a request to show terms of service only within 15 seconds of transitioning to
+        // the logged in state with the showTermsOfService bit set. This prevents forcing all users
+        // through the terms of service flow when enabled until they next login or refresh the
+        // page.
+        return this.props.showTermsOfService &&
```

Comment:
Is there any way to avoid coupling this to a timestamp? I'm assuming we're lacking a tri-state to distinguish not knowing the TOS requirement from knowing it to be true or false?